Allow Preloader to group through association queries

jhawthorn opened the pull request:
In https://github.com/rails/rails/pull/41385 and https://github.com/rails/rails/pull/41496 we added the ability for the Preloader to group queries. However we limited this to the simpler non-"through" associations.

This PR adds the ability to group queries from "through" associations (both for the middle and target records) in the same way. They can be grouped with other through associations or plain non-"through" associations.

To achieve this we've introduced a private `Preloader::Branch` class, which most of the logic previously in `Preloader` has been moved to. `Branch` represents a single association "branch" taken in the preloader (any key or value passed in by the user). This was a helpful concept to add because Branches with a "through" association needs 2 queries to be loaded and so needs 2 iterations of query batching to be run. `Branch` only refers to one branch/association, so it's no problem to keep around longer, `Preloader` can refer to multiple, so it would have required all queries to finish before any child_preloaders could be started.

We've added two tests, each of which now runs one less query.

Co-authored with @dinahshi 

I have hopes that this `Branch` class opens some future possibilities like choosing a more efficient order to load these associations (grouping queries at different "levels" of the hash) or running queries in parallel.

cc @eileencodes @HParker @byroot @rafaelfranca

jhawthorn committed:
Add ensure_loaded to Preloader::Association

This allows running either load_records or fetch_from_preloaded_records
depending on whether associations are already loaded. This allows
calling preloaded_records without necessarily having called run first.

Co-authored-by: Dinah Shi <dinahshi@github.com>

jhawthorn committed:
Allow preload grouping of through associations

Co-authored-by: Dinah Shi <dinahshi@github.com>

jhawthorn committed:
Implement Preloader::Branch

Co-authored-by: Dinah Shi <dinahshi@github.com>

casperisfine commented:
I tested this branch on our CI, and we have one test failing because there's now two queries instead of one. I'll try to dig that up tomorrow, but just wanted to let you know.

jhawthorn commented:
Thanks! This definitely should be reducing queries not increasing them 😓

casperisfine commented:
> This definitely should be reducing queries not increasing them 😓

Ahah, well, it's just one test among many, not that bad.

I'm working on writing a repro in Active Record's test suite, but for now I reduced it to:

```ruby
      ActiveRecord::Associations::Preloader.new(records: [@order], associations: [
        line_items: { line_items_applied_discounts: :discount_application },
        shipping_lines: { shipping_lines_applied_discounts: :discount_application },
      ]).call
```

I'll try to turn this into a test case in Active Record.

casperisfine commented:
Seems like I can't repro the double query in isolation... there must be something more to it, I'll have to dig more.

byroot committed:
Reproduce the preloader regression in #41596

It's very peculiar, what seem to happen is that if some associations
are already loaded, the batching is done properly.

casperisfine commented:
Ok, I think I found a repro and a fix: https://github.com/rails/rails/compare/main...Shopify:preloader_branch_testing

casperisfine commented on activerecord/lib/active_record/associations/preloader/branch.rb:
```diff
@@ -0,0 +1,113 @@
+        end
+
+        def children_hash
+          @children.map do |child|
+            [child.association, child.children_hash]
+          end.to_h
```
Nitpick:
```suggestion
          @children.to_h { |child| [child.association, child.children_hash] }
```

jhawthorn replied:
oops. I think I can delete this entire method

jhawthorn commented:
> Ok, I think I found a repro and a fix: [main...Shopify:preloader_branch_testing](https://github.com/rails/rails/compare/main...Shopify:preloader_branch_testing)

Aha! Thanks for the test that's super helpful to understand.

I'd argue that this isn't exactly a bug. The previous batching implementation would traverse the preloads breadth-first from the root of the hash passed in. After this PR it traverses breadth first from the first unloaded associations it finds. I'd argue both approaches are valid and there's merits to each.

That said, I don't want to make change unnecessarily. <y entire hope of introducing the Branch model is that soon we'll be able to make this choice more intelligently. So next week I'll change this to use the order it did before (fixing the regression test) and save that change for a future commit (where it should not create an extra query in this case).
